Simple Websocket & HTTP client API support

jukkar opened the pull request:
This PR provides a simple Websocket client API to send Websocket messages to Websocket server. Because the Websocket is established using HTTP, there is small and simple HTTP client API implemented. This HTTP client API can be used to connect to HTTP(S) servers so it is not tied to Websocket client implementation. As this HTTP API is very much needed by Websocket API, I decided not to send it separately to review.

The Websocket API provides functions that can be used by BSD socket API or if more fine grained Websocket support is needed, a Websocket specific API functions are provided.

The main function is this one
```
int websocket_connect(int http_sock, const char *host, const char *url,
		      const char **optional_headers,
		      websocket_connect_cb_t cb,
		      const struct http_parser_settings *http_cb,
		      u8_t *tmp_buf, size_t tmp_buf_len,
		      s32_t timeout, void *user_data);
```
It will be given a socket descriptor to the Websocket server. This means that all the non-websocket related things are done outside of this API. The `websocket_connect()` will connect to the server, do needed HTTP handshakes and return a new websocket socket descriptor that, when used, will add Websocket header to the sent data. User can then either use normal `send()` or `websocket_send_msg()` functions to send the data.

One usecase for this API is run MQTT data on top of it. This MQTT support is not part of this PR but to be done separately.

jukkar committed:
net: http: client: Initial version

Simple HTTP client API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: http_client: HTTP client sample application

Simple HTTP client sample that connects to HTTP server and does
GET and POST requests.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: websocket: client: Simple API for Websocket client

Implement simple API to do Websocket client requests.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: websocket: App for doing Websocket client requests

This is BSD sockets based application for connecting to
Websocket server.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print info about websocket

Add "net websocket" command that displays websocket information.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Verified that http-client sample works with TLS. Added README files to both samples.

rlubos commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,855 @@
+#include <net/http_client.h>
+#include <net/websocket.h>
+
+#include <sys/byteorder.h>
```
This was already included above.

rlubos commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,855 @@
+	struct websocket_context *ctx = NULL;
+	int i;
+
+	k_sem_take(&contexts_lock, K_FOREVER);
```
Any particular reason to use semaphore here instead of mutex?

jukkar replied:
No particular reason for that, just re-using older code that used semaphore.

rlubos commented on include/net/websocket.h:
```diff
@@ -0,0 +1,167 @@
+ * @details The function will automatically add websocket header to the
+ * message.
+ *
+ * @param ws_sock websocket id returned by websocket_connect(). Note that this
```
`Note that this socket is used to do HTTP handshakes etc. The actual Websocket connectivity is done via the returned websocket id.` - this sentence is not correct for this function (copy-paste from above)?

rlubos reviewed (commented):
Gone through most of the code, it looks that a lot of work was put into this!

jukkar commented on subsys/net/lib/http/http_client.c:
```diff
@@ -0,0 +1,616 @@
+	if (end_of_send > (int)send_buf_max_len) {
+		NET_ERR("Sending overflow (%d > %zd)", end_of_send,
+			send_buf_max_len);
+		return -EMSGSIZE;
```
Yes, the send buffer should not overflow. This check is just an extra check to make sure we don't overflow.

jukkar commented on subsys/net/lib/http/http_client.c:
```diff
@@ -0,0 +1,616 @@
+		req->response.http_cb->on_message_begin(parser);
+	}
+
+	if (IS_ENABLED(CONFIG_NET_HTTP_LOG_LEVEL_DBG)) {
```
Yep, NET_DBG() is enough here. This function had some extra stuff earlier but after removing that the outer if() can be removed too.

jukkar commented on subsys/net/lib/http/http_client.c:
```diff
@@ -0,0 +1,616 @@
+		total_received += received;
+		offset += received;
+
+		if (offset > req->response.recv_buf_len) {
```
Indeed, the check should be `>=`. Depending on the HTTP parsing state, it might be ok to start from the beginning of the buffer.

jukkar commented on samples/net/sockets/http_client/README.rst:
```diff
@@ -0,0 +1,91 @@
+====================
+
+Enable TLS support in the sample by building the project with the
+``overlay-tls.conf`` overlay file enabled (found in echo_server directory),
```
> perhaps it's a better idea to provide own `overlay-tls.conf` for this sample

Sure, makes sense.

jukkar commented:
Updated according to comments.

dbkinder commented on samples/net/sockets/http_client/README.rst:
```diff
@@ -0,0 +1,90 @@
+Running http-server in Linux Host
+=================================
+
+There is one useful testing scenario that can be used with Linux host.
```
How about squishing this into one sentence:
```
You can run this  `http-client`` sample application in QEMU
and run the ``http-server.py`` (from net-tools) on a Linux host.
```

dbkinder reviewed (changes requested):
some suggested edits...

jukkar commented:
> some suggested edits...

Thanks David. New updated version uploaded.

jukkar commented:
Updated documentation in websocket sample. Fixed case when we received two websocket packets in same TCP segment, in that case we need to remember the second websocket msg for next socket read.

jukkar commented:
ping @dbkinder can you re-check the modifications

dbkinder reviewed (approved):
+1 for doc changes

jukkar commented:
Fixed the handling of Sec-Websocket-Accept field in HTTP header.

rlubos reviewed (approved):
Thanks

pfalcon commented:
Sorry, I was on vacation, then at the conference, so late to the party. Commenting first just based on the textual description and comments here.

> The main function is this one

>int websocket_connect(**WALL OF PARAMETERS**);

Didn't we discussed once that having gigantic parameter list is *worse* than a dedicated structure to capture those params instead? Some params are strange to, e.g.:

> 		      websocket_connect_cb_t cb,

"sockets" and "callbacks" are generally don't mix together. How events are reported for sockets are via pull-style poll() call (instead of push-style callback). If you (for some reason) try to represent high-level websocket communication channel as a system socket, so callbacks, again, look strange. 

> 		      const struct http_parser_settings *http_cb,

Why structure of type "http_parser_settings" is named as if it was callback?

> const char **optional_headers,

What if I don't have all the headers preallocated as single static array (e.g., don't have enough memory for that)?

> all the non-websocket related things are done outside of this API. The websocket_connect() will connect to the server, do needed HTTP handshakes 

The first part seems to contradict the second. So, does it issue connect() call? It shouldn't. And if doesn't, then I'd skip mentioning that it "will connect to the server". Mentioning "do needed HTTP handshakes" should be enough, perhaps elaborating it to "will do needed HTTP/Websocket handshake on already established connection".

> and return a new websocket socket descriptor that,

If it returns *file descriptor*, that's our old good friend layering violation. Why conjugate high-level websocket connection to pretend it's a system-level socket. No other (sane) system does that. There're immediate portability implications, e.g. you can't test this websocket code on Linux with comfort it gives, so it's likely has bugs. And if it doesn't try to represent itself as a system-level socket, why return an int, instead of structure pointer? 

> when used, will add Websocket header to the sent data. User can then either use normal send() or websocket_send_msg() functions to send the data.

Hey, that's cool! But what about received data? ;-)

pfalcon commented:
> Fixed case when we received two websocket packets in same TCP segment, in that case we need to remember the second websocket msg for next socket read.

Websocket is SOCK_STREAM protocol, and SOCK_STREAM sockets don't work in terms of "TCP segments", they work in terms of "streams of bytes". So, if BSD Socket API was properly used to implement websocket code, this error couldn't happen in first place... I'm now a bit afraid to look into the actual code...

jukkar commented on include/net/http_client.h:
```diff
@@ -0,0 +1,240 @@
+ * @{
+ */
+
+#include <kernel.h>
```
HTTP is implemented on top of socket as you probably noticed, so did not got your comment. The kernel.h can be removed, dunno why it was there in the first place.

pfalcon replied:
+1 for removing it. My point is that, if you can take this code, and compile it for Linux, then it's an ultimately best situation, proving code portability and quality (if we can test it on Linux too).

jukkar commented on include/net/http_client.h:
```diff
@@ -0,0 +1,240 @@
+	 */
+	char http_status[HTTP_STATUS_STR_SIZE];
+
+	u8_t cl_present:1;
```
sure

jukkar commented on include/net/http_client.h:
```diff
@@ -0,0 +1,240 @@
+	 * some header fields may remain constant through the application's
+	 * life cycle. This is a NULL terminated list of header fields.
+	 */
+	const char **header_fields;
```
You create a HTTP request by setting things up in your application, then calling the API func. What do you mean by generating headers dynamically in this case, what kind of dynamical behaviour your application would need?

pfalcon replied:
> What do you mean by generating headers dynamically in this case, what kind of dynamical behaviour your application would need?

Well, suppose an app wants to send 2 headers with a payload of 2048 bytes for each (some auth-related material for example). Then, giving an option to send headers one by one would help. This could be achieved by providing flexible low-level API, allowing user to construct request piece by piece, e.g. request_start(), request_send_header(hdr, value) (can be called multiple times), request_send_payload() (can be called multiple times), request_end(). On top of that, "convenience" functions can be added for most common scenarios.


And it's just an idea I wanted to mention in review comments, I don't call it to be reworked like that so late in the review cycle. (Well, unless you think otherwise ;-) ).

jukkar commented on samples/net/sockets/http_client/src/main.c:
```diff
@@ -0,0 +1,351 @@
+{
+	const char *content[] = {
+		"foobar",
+		"chuncked",
```
typo, will fix it, anyway it is part of the HTTP payload so it does not really matter what combination of chars we send

pfalcon replied:
> anyway it is part of the HTTP payload so it does not really matter what combination of chars we send

That's true, but still, unless typo is made on purpose, would be nice to keep the code grammar-correct (or otherwise it may lead to a feeling of sloppiness) ;-).

jukkar commented:
> Websocket is SOCK_STREAM protocol, and SOCK_STREAM sockets don't work in terms of "TCP segments", they work in terms of "streams of bytes". So, if BSD Socket API was properly used to implement websocket code, this error couldn't happen in first place... I'm now a bit afraid to look into the actual code...

The Websocket is a message protocol on top of stream protocol (TCP). IMHO it is not very well designed in the first place but we need to cope with this.

jukkar commented:
> Didn't we discussed once that having gigantic parameter list is _worse_ than a dedicated structure to capture those params instead? Some params are strange to, e.g.:

Do you have a proposal how to pass the params?

> Why structure of type "http_parser_settings" is named as if it was callback?

The struct contains HTTP callback pointers.

> What if I don't have all the headers preallocated as single static array (e.g., don't have enough memory for that)?

That is not possible with this API. We would probably need to have a callback to feed extra headers to the stream.

> > all the non-websocket related things are done outside of this API. The websocket_connect() will connect to the server, do needed HTTP handshakes
> 
> The first part seems to contradict the second. So, does it issue connect() call? It shouldn't. And if doesn't, then I'd skip mentioning that it "will connect to the server". Mentioning "do needed HTTP handshakes" should be enough, perhaps elaborating it to "will do needed HTTP/Websocket handshake on already established connection".

There is no contradiction here, user is responsible to call connect(), the websocket API will then *connect* i.e., establish, like doing handshakes etc., a valid connection to websocket server.

> > and return a new websocket socket descriptor that,
> 
> If it returns _file descriptor_, that's our old good friend layering violation. Why conjugate high-level websocket connection to pretend it's a system-level socket. No other (sane) system does that. There're immediate portability implications, e.g. you can't test this websocket code on Linux with comfort it gives, so it's likely has bugs. And if it doesn't try to represent itself as a system-level socket, why return an int, instead of structure pointer?

We want to use socket descriptor that is tied to websocket connection. This helps to run things on top of websocket connection and use recv()/send() API. We are not making Linux applications here, and if you want to run valgrind etc you can use native_posix for that.

> > when used, will add Websocket header to the sent data. User can then either use normal send() or websocket_send_msg() functions to send the data.
> 
> Hey, that's cool! But what about received data? ;-)

recv() works just fine too

jukkar commented:
Changing slightly the API proposal:
* moved most of the function parameters to user given request struct
* typos fixed

pfalcon commented:
> Do you have a proposal how to pass the params?

Not something specific, beyond moving all (most) params to a struct, and passing a pointer to it to the function. I see that you already pushed something along those lines, let me have a look.

pfalcon commented on include/net/http_client.h:
```diff
@@ -0,0 +1,289 @@
+ * @return <0 if error, >=0 amount of data sent to the server
+ */
+int http_client_req(int sock, struct http_request *req,
+		    s32_t timeout, void *user_data);
```
Is there a reason why `user_data` is passed as a separate param and is not part of `struct http_request`? (Just in case, I'm asking, not calling for change.)

jukkar replied:
Many APIs in networking side etc. have typically timeout and user_data parameters in function calls. Just followed this style here.

pfalcon replied:
> Just followed this style here.

Makes sense.

pfalcon reviewed (commented):
I've reviewed this PR, and agree that it's useful and good deal of effort went into it. I appreciate reacting to some suggestions I've made. I have to say that I'm still not fully satisfied with it due to the fact that HTTP API uses callbacks instead of pull-style, socket-style read()/write() functions (as proposed in https://github.com/zephyrproject-rtos/zephyr/issues/13212). So, I'd give this +0.4, which rounds to +0.

At the same time, I understand why it's done like that - callbacks ultimately come from 3rd-party HTTP protocol parser we use, and "masking" them would just take extra effort, especially if the real task here was implementing Websocket support. The concern with callbacks is their known composability issues, but this very PR somehow counter-proves that, as it was possible to implement socket-style API for websocket on top of the underlying callback-style HTTP API (though it's unclear what issues may be lurking there) . So, overall, I don't think I'd come up with an overall better initial implementation even if I tried a callback-less route.

I definitely hope this will be merged soonish, to allow more people to play with it. In that regard, if my approval would help to merge it sooner rather than later, I'd be happy to give my +1 just for that reason, even though I'm not fully satisfied with the overall design.

jukkar commented on include/net/websocket.h:
```diff
@@ -0,0 +1,202 @@
+
+	/** User supplied callback function to call when optional headers need
+	 * to be sent. This can be NULL, in which case the optional_headers
+	 * field in http_request is used. The idea of this payload callback is
```
Note to self, typo s/payload/optional_headers/

jukkar commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,949 @@
+	msg.msg_iov = io_vector;
+	msg.msg_iovlen = ARRAY_SIZE(io_vector);
+
+#if HEXDUMP_SENT_PACKETS
```
need to replace this with `if (HEXDUMP_SENT_PACKETS)`

jukkar commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,949 @@
+			pos_to_write = ctx->pos - header_len;
+		}
+
+#if HEXDUMP_RECV_PACKETS
```
need to replace this with `if (HEXDUMP_RECV_PACKETS)`

jukkar commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,949 @@
+	struct websocket_context *ctx = obj;
+	s32_t timeout = K_FOREVER;
+
+	if ((flags & ZSOCK_MSG_DONTWAIT)) {
```
extra ()

jukkar commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,949 @@
+	struct websocket_context *ctx = obj;
+	s32_t timeout = K_FOREVER;
+
+	if ((flags & ZSOCK_MSG_DONTWAIT)) {
```
here too

jukkar reviewed (commented):
Noticed some typos and minor nits when doing self review. I will fix these.

jukkar commented:
Noticed some fd leaks when closing the connection, those are now fixed.

jukkar commented:
Seems to work ok now. I have been running mqtt_publisher over Websocket without any issues.
